com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java

```diff
@@ -108,6 +112,28 @@ public void initialize(IDebugAdapterContext context, Map<String, Object> props)
         return evaluate(expression, thread, depth, null);
     }
 
+    @Override
+    public CompletableFuture<Value> evaluate(String expression, ObjectReference context, ThreadReference thread, int depth) {
+        CompletableFuture<Value> completableFuture = new CompletableFuture<>();
+        try  {
+            ensureDebugTarget(thread.virtualMachine(), thread, depth);
+            JDIThread jdiThread = getMockJDIThread(thread);
+            JDIStackFrame stackframe = createStackFrame(jdiThread, depth);
+            if (stackframe == null) {
+                throw new IllegalStateException("Cannot evaluate because the stackframe is not available.");
+            }
+
+            JDIObjectValue jdiObject = new JDIObjectValue(debugTarget, context);
+            ASTEvaluationEngine engine = new ASTEvaluationEngine(project, debugTarget);
+            ICompiledExpression compiledExpression = engine.getCompiledExpression(expression, jdiObject);
```

Comment:
for conditional breakpoints, the compiled expression is saved into breakpoint for better performance, convert evaluation text to compiled expression is heavy, be aware of that and take stress tests on your code.

Reply:
The variable evaluation is requested when the user click the complex object to expand its children. And a single click only triggers one or two evaluation action. No need of stress test.